src/garage/tf/policies/base.py

```diff
@@ -23,41 +23,68 @@ def __init__(self, name, env_spec):
         self._cached_param_shapes = None
 
     @abc.abstractmethod
-    def get_action(self, observation):
+    def get_action(self, observation, agent_infos):
         """Get action sampled from the policy.
 
         Args:
             observation (np.ndarray): Observation from the environment.
+            agent_infos (dict): Infos for previous action and hidden states.
 
         Returns:
             (np.ndarray): Action sampled from the policy.
 
         """
 
     @abc.abstractmethod
-    def get_actions(self, observations):
+    def get_actions(self, observations, agent_infos):
         """Get action sampled from the policy.
 
         Args:
             observations (list[np.ndarray]): Observations from the environment.
+            agent_infos (dict): Infos for previous action and hidden states.
 
         Returns:
             (np.ndarray): Actions sampled from the policy.
 
         """
 
-    def reset(self, dones=None):
+    def reset(self, agent_infos, dones=None):
         """Reset the policy.
 
         If dones is None, it will be by default np.array([True]) which implies
         the policy will not be "vectorized", i.e. number of parallel
         environments for training data sampling = 1.
 
         Args:
+            agent_infos (dict): Infos for previous action and hidden states.
             dones (numpy.ndarray): Bool that indicates terminal state(s).
 
         """
 
+    def get_initial_state(self, dones=None):
+        """Initial state.
+
+        Note:
+            If `dones` is None, it will be by default `np.array([True])` which
+            implies the policy will not be "vectorized", i.e. number of
```

Comment:
hmm i think that instead `get_action` should also be stateless with respect to the vectorization. it seems here that the shape of `dones` implies that `get_actions` will be called only with shape `(P)` in the future, but I think this is an artificial constraint.

it's pretty trivial to require vectorized users to tile `initial_state` into a shape which matches `observations`.

a consistent way to implement this to add a function `get_initial_states` to this interface and consider that the default way it will be used. the relationship between `get_initial_state` and `get_initial_states` is the same as the relationship between `get_action` and `get_actions`: `get_initial_states` is fully-vectorized and returns the initial state tensor given some prototype array to determine the batch size, `get_initial_state` can always be implemented as something like `self.get_initial_states(batch_size)[0]`

Reply:
`...given some prototype array to determine the batch size...`
How about directly passing the batch size? like
```python
    def get_initial_state(self):
        return self.get_initial_states(batch_size=1)

    def get_initial_states(self, batch_size):
        initial_hidden = self.model.networks['default'].init_hidden.eval()
        prev_actions = np.zeros((batch_size, self.action_space.flat_dim))
        prev_hiddens = np.full((batch_size, self._hidden_dim), initial_hidden)

        return dict(prev_action=prev_actions, prev_hidden=prev_hiddens)
```

Reply:
that seems reasonable to me. are you sure about the implementation of `get_initial_state`? don't you need `self.get_initial_states(batch_size=1)[0]` or `np.squeeze(self.get_initial_states(batch_size=1)`?

Reply:
I think I need to do `{k: v[0] for k, v in state.items()}`